File: client/js/box.js
// Implements simple collision detection between the world
// and a box whose center is at p and has a size described
// by the given halfExtents. Does not do partial application
// of movements. Feel free to improve it!
function Box(p, halfExtents, centerOffset) {
	var self = this;
	centerOffset = centerOffset || new THREE.Vector3(0.0, 0.0, 0.0);

	self.setPos = function (newPos) {
		p = newPos;
	};

	self.attemptMove = function (world, move) {
		if (inSolid(world)) {
			var gh = groundHeight(world);
			move.y = p.y - gh;
			p.y = gh;
		}

		p.x += move.x;
		if (inSolid(world)) {
			p.x -= move.x;
			move.x = 0;
		}

		p.y += move.y;
		if (inSolid(world)) {
			p.y -= move.y;
			move.y = 0;
		}

		p.z += move.z;
		if (inSolid(world)) {
			p.z -= move.z;
			move.z = 0;
		}

		return p;
	};

	self.contains = function (x, y, z) {
		var he = halfExtents;
		var co = centerOffset;
		var xs = p.x + co.x - he.x, xe = p.x + co.x + he.x;
		var ys = p.y + co.y - he.y, ye = p.y + co.y + he.y;
		var zs = p.z + co.z - he.z, ze = p.z + co.z + he.z;

		return xs < x && xe > x &&
			ys < y && ye > y &&
			zs < z && ze > z;
	};

	function bboxEach(fn, reduce) {
		var he = halfExtents;
		var co = centerOffset;
		var xs = p.x + co.x - he.x, xe = p.x + co.x + he.x;
		var ys = p.y + co.y - he.y, ye = p.y + co.y + he.y;
		var zs = p.z + co.z - he.z, ze = p.z + co.z + he.z;
		// TODO: Figure out where more points are needed
		// based on the size of the shape (we use this many
		// because this is needed given the current size of
		// our player model).
		var ym = ys + (ye - ys) / 2;
		return reduce(
			fn(xs, ys, zs),
			fn(xs, ys, ze),
			fn(xs, ye, zs),
			fn(xs, ye, ze),
			fn(xs, ym, zs),
			fn(xs, ym, ze),
			fn(xe, ys, zs),
			fn(xe, ys, ze),
			fn(xe, ye, zs),
			fn(xe, ye, ze),
			fn(xe, ym, zs),
			fn(xs, ym, ze)
		);
	}

	function logicalOr() {
		for (var i = 0; i < arguments.length; i++) {
			if (arguments[i]) return true;
		}
		return false;
	}

	function inSolid(world) {
		function solidHere(x, y, z) {
			var block = world.blockAt(x, y, z);
			if (!block) return true;
			else return block.solid();
		}
		return bboxEach(solidHere, logicalOr);
	}

	function groundHeight(world) {
		var cg = world.findClosestGround;
		return bboxEach(cg, Math.max) + halfExtents.y - centerOffset.y;
	}
}

File: server/src/buildblast/physics/box.go
package physics

import (
	"buildblast/coords"
	"buildblast/mapgen"
)

type Box struct {
	xs float64
	xe float64
	ys float64
	ym float64
	ye float64
	zs float64
	ze float64
}

func NewBox(position coords.World, halfExtents coords.Vec3) *Box {
	return NewBoxOffset(position, halfExtents, coords.Vec3{X: 0, Y: 0, Z: 0})
}

func NewBoxOffset(position coords.World, halfExtents coords.Vec3, centerOffset coords.Vec3) *Box {
	p := position
	he := halfExtents
	co := centerOffset
	return &Box{
		xs: p.X - he.X + co.X,
		xe: p.X + he.X + co.X,
		ys: p.Y - he.Y + co.Y,
		ym: p.Y + co.Y,
		ye: p.Y + he.Y + co.Y,
		zs: p.Z - he.Z + co.Z,
		ze: p.Z + he.Z + co.Z,
	}
}

func (b *Box) AttemptMove(world mapgen.BlockSource, amount coords.Vec3) coords.Vec3 {
	if b.inSolid(world) {
		return amount
	}

	b.xs += amount.X
	b.xe += amount.X
	if (b.inSolid(world)) {
		b.xs -= amount.X
		b.xe -= amount.X
		amount.X = 0
	}

	b.ys += amount.Y
	b.ym += amount.Y
	b.ye += amount.Y
	if (b.inSolid(world)) {
		b.ys -= amount.Y
		b.ym -= amount.Y
		b.ye -= amount.Y
		amount.Y = 0
	}

	b.zs += amount.Z
	b.ze += amount.Z
	if (b.inSolid(world)) {
		b.zs -= amount.Z
		b.ym -= amount.Z
		b.ye -= amount.Z
		amount.Z = 0
	}

	return amount;
}

func (b *Box) Contains(position coords.World) bool {
	panic("Contains() not implemented!")
}

func (b *Box) inSolid(world mapgen.BlockSource) bool {
	solid := func (x, y, z float64) bool {
		block := world.Block(coords.World{x, y, z})
		return block.Solid()
	}
	xs := b.xs; xe := b.xe; ys := b.ys; ym := b.ym
	ye := b.ye; zs := b.zs; ze := b.ze
	return solid(xs, ys, zs) || solid(xs, ys, ze) ||
		   solid(xs, ym, zs) || solid(xs, ym, ze) ||
		   solid(xs, ye, zs) || solid(xs, ye, ze) ||
		   solid(xe, ys, zs) || solid(xe, ys, ze) ||
		   solid(xe, ym, zs) || solid(xe, ym, ze) ||
		   solid(xe, ye, zs) || solid(xe, ye, ze)
}
